runtime/dart_isolate.cc

```diff
@@ -280,8 +283,159 @@ std::weak_ptr<DartIsolate> DartIsolate::CreateRootIsolate(
+    // because the engine is shutting down.
+    FML_LOG(INFO) << "CreatePlatformIsolate called after shutdown";
+    return nullptr;
+  }
+
+  Dart_Isolate parent_isolate = isolate();
+  Dart_ExitIsolate();  // Exit parent_isolate.
+
+  const TaskRunners& task_runners = GetTaskRunners();
+  fml::RefPtr<fml::TaskRunner> platform_task_runner =
+      task_runners.GetPlatformTaskRunner();
+  FML_DCHECK(platform_task_runner);
+
+  auto isolate_group_data = std::shared_ptr<DartIsolateGroupData>(
+      *static_cast<std::shared_ptr<DartIsolateGroupData>*>(
+          Dart_IsolateGroupData(parent_isolate)));
+
+  Settings settings(isolate_group_data->GetSettings());
+
+  // PlatformIsolate.spawn should behave like Isolate.spawn when unhandled
+  // exceptions happen (log the exception, but don't terminate the app). But the
+  // default unhandled_exception_callback may terminate the app, because it is
+  // only called for the root isolate (child isolates are managed by the VM and
+  // have a different error code path). So override it to simply log the error.
+  settings.unhandled_exception_callback = [](const std::string& error,
+                                             const std::string& stack_trace) {
+    FML_LOG(ERROR) << "Unhandled exception:\n" << error << "\n" << stack_trace;
+    return true;
+  };
+
+  // The platform isolate task observer must be added on the platform thread. So
+  // schedule the add function on the platform task runner.
+  TaskObserverAdd old_task_observer_add = settings.task_observer_add;
+  settings.task_observer_add = [old_task_observer_add, platform_task_runner,
+                                platform_isolate_manager](
+                                   intptr_t key, const fml::closure& callback) {
+    platform_task_runner->PostTask([old_task_observer_add,
+                                    platform_isolate_manager, key, callback]() {
+      if (platform_isolate_manager->IsShutdown()) {
+        // Shutdown happened in between this task being posted, and it running.
+        // platform_isolate has already been shut down. Do nothing.
+        FML_LOG(INFO) << "Shutdown before platform isolate task observer added";
+        return;
+      }
+      old_task_observer_add(key, callback);
+    });
+  };
+
+  UIDartState::Context context(task_runners);
+  context.advisory_script_uri = isolate_group_data->GetAdvisoryScriptURI();
+  context.advisory_script_entrypoint =
+      isolate_group_data->GetAdvisoryScriptEntrypoint();
+  auto isolate_data = std::make_unique<std::shared_ptr<DartIsolate>>(
+      std::shared_ptr<DartIsolate>(new DartIsolate(settings,  // settings
+                                                   false,     // is_root_isolate
+                                                   true,  // is_platform_isolate
+                                                   context)));  // context
+  (*isolate_data)->platform_isolate_manager_ = platform_isolate_manager;
```

Comment:
Add a new `DartIsolate` constructor that takes the `PlatformIsolateManager` as an argument and sets the `is_platform_isolate_` flag.

This avoids the need to access the private `platform_isolate_manager_` field here and allows removal of `is_platform_isolate` from the original `DartIsolate` constructor.